src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+          maxBaseQuality   = options.maxBaseQuality,
+          baseQualityShift = options.baseQualityShift,
+          errorRatePostUmi = options.errorRatePostUmi
+        )
+      } // NB: binary Phred, not ASCII
+
+    // get the most likely consensus bases and qualities
+    val (consensusBases, consensusQualities) = consensusCalls(
+      bases                   = bases,
+      quals                   = quals,
+      errorRatePreUmi         = options.errorRatePreUmi,
+      minConsensusBaseQuality = options.minConsensusBaseQuality,
+      minReads                = options.minReads
+      )
+
+    // check that the mean base quality is high enough
+    if (PhredValue.mean(consensusQualities:_*) < options.minMeanConsensusBaseQuality) None
+    else Some((consensusBases, SAMUtils.phredToFastq(consensusQualities.map(_.toByte).toArray[Byte])))
+  }
+
+  /** Creates a consensus read from the given records.  If no consensus read was created, None is returned. */
+  def fromSamRecords(records: Seq[SAMRecord],
+                     options: ConsensusCallerOptions = new ConsensusCallerOptions()
+                    ): Option[(String, String)] = {
+    // TODO: if mapped, this gets the bases in genomic order and not sequencing order.  Does it matter?
+    val bases = records.map { rec => rec.getReadString.toUpperCase }
+    val quals = records.map { rec => rec.getBaseQualityString }
+
+    fromBasesAndQualities( basesAndQualities=bases.zip(quals), options = options)
+  }
+
+  /** Get the most likely consensus bases and qualities. */
+  private[bam] def consensusCalls(bases: Seq[String],
+                                  quals: Seq[Seq[PhredValue]],
+                                  errorRatePreUmi: PhredValue         = DefaultErrorRatePreUmi,
+                                  minConsensusBaseQuality: PhredValue = DefaultMinConsensusBaseQuality,
+                                  minReads: Int                       = DefaultMinReads): (String, Seq[PhredValue]) = {
+    val maxBaseLength = bases.map(_.length).max
+
+    // go through each position in the read(s).
+    val (consensusBases, consensusQualities) = Seq.range(0, maxBaseLength, 1).map { baseIdx =>
+      val basesAtIdx = bases.filter { str => baseIdx < str.length}.map { str => str.charAt(baseIdx) }
+
+      // check to see if we heave a enough reads to produce a consensus base.
+      if (basesAtIdx.length < minReads) (SequenceUtil.N.toChar, ZeroPhred)
+      else {
+        val qualsAtIdx         = quals.filter { q => baseIdx < q.length}.map { q => q(baseIdx) }
+        val basesAndQualsAtIdx = basesAtIdx.zip(qualsAtIdx)
+
+        // TODO: handle Ns in the reads
+        // get the phred-scaled probability of each potential DNA base
+        var phreds = SequenceUtil.VALID_BASES_UPPER.map(_.toChar).map {
+          candidateBase =>
+            basesAndQualsAtIdx.map { case (base, qual) =>
+              if (base == candidateBase) qual.inv() // probability of not an error
```

Comment:
Something doesn't sit right with me here.  You use `qual.inv()`, which I would interpret as `1 / qual`.  But other operators on `PhredValue` are named as if dealing with non-log values (e.g. `phred * phred` is multiplication in linear space, and implemented as summing in log space).  Should `inv()` therefore be `oneMinus()`?